dartagnan/src/main/java/com/dat3m/dartagnan/program/analysis/BackwardsReachingDefinitionsAnalysis.java

```diff
@@ -0,0 +1,384 @@
+
+        private ReaderInfo(Set<Register> u) {
+            used = Set.copyOf(u);
+        }
+
+        @Override
+        public Set<Register> getUsedRegisters() {
+            return used;
+        }
+
+        @Override
+        public RegisterWriters ofRegister(Register register) {
+            return new RegisterWritersView(this, register);
+        }
+    }
+
+    private static final class RegisterWritersView implements RegisterWriters {
+
+        private final ReaderInfo info;
+        private final Register register;
+
+        private RegisterWritersView(ReaderInfo info, Register register) {
+            this.info = info;
+            this.register = register;
+        }
+
+        @Override
+        public List<RegWriter> getMayWriters() {
+            return info.mayWriters.stream().filter(w -> w.getResultRegister().equals(register)).toList();
+        }
+
+        @Override
+        public List<RegWriter> getMustWriters() {
+            return info.mustWriters.stream().filter(w -> w.getResultRegister().equals(register)).toList();
+        }
+
+        @Override
+        public boolean mustBeInitialized() {
+            return !info.uninitialized.contains(register);
+        }
+    }
+
+    /**
+     * Local information for either a {@link RegWriter} or the initial state of the program.
+     */
+    public static final class Readers {
+
+        private final Set<RegReader> readers = new HashSet<>();
+        private boolean mayBeFinal;
+
+        private Readers() {}
+
+        /**
+         * Lists readers s.t. there may be an execution where an instance of that reader uses the result of an instance of the associated writer.
+         * @return Unordered set of Readers.
+         */
+        public Collection<RegReader> getReaders() {
+            return Collections.unmodifiableCollection(readers);
+        }
+
+        /**
+         * @return {@code true} if there may be an execution, where an instance of the associated writer defines the last value of its register.
+         */
+        public boolean mayBeFinal() {
+            return mayBeFinal;
+        }
+    }
+
+    private BackwardsReachingDefinitionsAnalysis() {}
+
+    private void initialize(Function function, Set<Register> finalRegisters) {
+        for (Event event : function.getEvents()) {
+            if (event instanceof RegWriter writer) {
+                writerMap.put(writer, new Readers());
+            }
+        }
+        writerMap.put(null, new Readers());
+        for (Event event : function.getEvents()) {
+            if (event instanceof RegReader reader) {
+                final Set<Register> usedRegisters = new HashSet<>();
+                for (Register.Read read : reader.getRegisterReads()) {
+                    usedRegisters.add(read.register());
+                }
+                readerMap.put(reader, new ReaderInfo(usedRegisters));
+            }
+        }
+        readerMap.put(null, new ReaderInfo(finalRegisters));
```

Comment:
Why not process everything in the same loop iteration? I don't see any dependency from the second loop tot eh first one.

Also, what are these `X.put(null, ...)` representing?

Reply:
The `null` entry is likely for reads from uninitialized registers.
I agree about the loop: either you merge them or you keep them and iterate over `getEvents(RegWriter.class)` resp. `getEvents(RegReader.class)` to skip the `instanceof` checks.

Reply:
> The `null` entry is likely for reads from uninitialized registers. 

Whatever this is used for, it should be documented, I don't want someone reading the code to have to guess

Reply:
Yep, since there are no dependencies, both versions are valid.  I consider my version better in terms of cache optimization, but not by much.

In `writerMap`, `null` stands for the initial writer(s).  In `readerMap`, it denotes the final reader(s).  Both cases don't have actual `Event` instances.  In this analysis, they behave *as if* they had.  For readability, I added named null constants.